File: FLUTTER/lib/Model/library.dart
import 'dart:async';

import 'package:firebase_database/firebase_database.dart';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:library_app/Model/book.dart';
import 'package:library_app/Model/cache_manager.dart';
import 'package:library_app/View/filter.dart';
import 'package:library_app/main.dart';

enum ResultType {
  paginated, plain
}

enum PaginationAction {
  getNextPage, getPreviousPage, stayAtCurentPage
}

class Library {
  late List<Book> _books;
  late DatabaseReference _dbLibraryReference;
  late Reference _storageReference;
  late CacheManager imagesCache;
  int currentPage;

  final int _nPageItems;
  final int _lastPage = 3;
  
  late ResultType resultType;
  late Filters? filters;

  int get booksCount => _books.length;
  int get pageSize => _nPageItems;

  Library(this._nPageItems, {this.currentPage = 1}){
    _dbLibraryReference = FirebaseDatabase
        .instance
        .refFromURL(dbUrlPath)
        .child("Library");

    _storageReference = FirebaseStorage
        .instance
        .refFromURL(sgUrlPath)
        .child("images");

    imagesCache = CacheManager(cacheUpdaterSource: _storageReference, nMinutesCacheTimeout: 30);

    _books = List.empty(growable: true);
    resultType = ResultType.paginated;
    filters = Filters(isWithoutReviews: false, fromPrice: "", toPrice: "");
  }

  Book? getBook({required int index}) {
    return _books.elementAtOrNull(index);
  }

  Future loadLibrary(PaginationAction action, {bool Function(Book)? filter}) async {
    if ((_lastPage > currentPage && action == PaginationAction.getNextPage) || 
        (currentPage > 1 && action == PaginationAction.getPreviousPage) || 
        action == PaginationAction.stayAtCurentPage) {    
      switch (action) {
        case PaginationAction.getNextPage: currentPage++;
        case PaginationAction.getPreviousPage: currentPage--;
        default:
      }
      
      Query query;
      resultType = filter == null ? ResultType.paginated : ResultType.plain;

      switch (resultType) {
        case ResultType.plain: query = _dbLibraryReference.orderByKey();
        case ResultType.paginated: query = _dbLibraryReference.orderByKey().startAt("${(currentPage - 1) * _nPageItems + 1}").limitToFirst(_nPageItems);
      }

      await query
      .get().then((value) async {     
        List<Book> tempArray = [];

        for (var jsonEncodedBook in value.children) {
          var book = Book.fromJson((jsonEncodedBook.value as Map).cast<String, dynamic>())..id = jsonEncodedBook.key!;
          if (filter != null) {
            if (filter(book)) tempArray.add(book);
          } else {
            tempArray.add(book);
          }

          book.images ??= List.empty(growable: true);
          book.images!.add(await imagesCache.get(name: book.imagesLinks[0]));      
          await loadBookImages(book);      
        }

        _books = tempArray;
      });

      return false;
    } else if (_lastPage == currentPage || currentPage == 0) {
      return true;
    }
  }

  Future loadBookImages(Book book, {Function? reloadState}) async {
    for (int i = 1; i < book.imagesLinks.length; i++) {
      book.images!.add(await imagesCache.get(name: book.imagesLinks[i]).whenComplete(() => reloadState));
    }
  }

  static Future updateBook(Book book) async{
    DatabaseReference dbBook = FirebaseDatabase.instance.refFromURL(dbUrlPath).child("Library");
    await dbBook.child(book.id).set(book.toJson());
  }
}

File: KOTLIN/src/main/java/by/bsuir/library/model/Library.kt
package by.bsuir.library.model

import android.content.Context
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.google.gson.GsonBuilder
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.tasks.await
import java.util.concurrent.CompletableFuture

const val dbUrlPath = "https://mpproject-750b8-default-rtdb.europe-west1.firebasedatabase.app/"
const val sgUrlPath = "gs://mpproject-750b8.appspot.com"
val filter = Filters(false, "", "")

class Library(private var context: Context, val onDataChanged: () -> Unit) {
    var books: MutableList<Book> = mutableListOf()
    private var dbLibraryReference: DatabaseReference = FirebaseDatabase
        .getInstance()
        .getReferenceFromUrl(dbUrlPath)
        .child("Library")
    private var storageReference: StorageReference = FirebaseStorage
        .getInstance()
        .getReferenceFromUrl(sgUrlPath)
        .child("images")
    private var imagesCache: CacheManager = CacheManager(context, storageReference, 30)
    private var filters: Filters? = null

    val booksCount: Int
        get() = books.size

    init {
        filters = Filters(isWithoutReviews = false, fromPrice = "", toPrice = "")
    }

    fun getBook(index: Int): Book? {
        return books.getOrNull(index)
    }

    fun loadLibrary(filter: ((Book) -> Boolean)? = null) {
        books.clear()
        dbLibraryReference.orderByKey().get().addOnSuccessListener {
            it.children.forEach { jsonEncodedBook ->
                val book = gson.fromJson(gson.toJson(jsonEncodedBook.value), Book::class.java)
                book.ID = jsonEncodedBook.key!!
                println(book.toString())

                if (filter == null || filter(book)) {
                    books.add(book)
                }

                book.images = mutableListOf()
                book.images!!.add(CompletableFuture.completedFuture(runBlocking { imagesCache.get(book.imagesLinks[0]) }).get())
                CompletableFuture.completedFuture(runBlocking { loadBookImages(book) }).get()
            }
            onDataChanged()
        }
    }

    private suspend fun loadBookImages(book: Book, reloadState: (() -> Unit)? = null) {
        book.images!!.clear()
        book.imagesLinks.forEach { imageLink ->
            book.images!!.add(imagesCache.get(imageLink).also { reloadState?.invoke() })
        }
    }

    companion object {
        private var gson = GsonBuilder().create()

        fun updateBook(book: Book) {
            val dbBook = FirebaseDatabase.getInstance().getReferenceFromUrl(dbUrlPath).child("Library")
            CompletableFuture.completedFuture(runBlocking { dbBook.child(book.ID).setValue(book.toJson()) }).get()
        }
    }
}


